packages/backend/server/src/core/utils/blocksuite.ts

```diff
@@ -0,0 +1,135 @@
+// TODO(@forehalo):
+//   Because of the `@affine/server` package can't import directly from workspace packages,
+//   this is a temprory solution to get the block suite data(title, description) from given yjs binary or yjs doc.
+//   The logic is mainly copied from packages/frontend/core/src/modules/docs-search/worker/in-worker.ts:crawlingDocData,
+//   and it's better to be provided by blocksuite
+
+import { Doc, Map } from 'yjs';
+
+export type PageDocContentType =
+  | {
+      type: 'affine:paragraph';
+      text: string;
+      variant: 'text' | 'h1' | 'h2' | 'h3' | 'h4';
+    }
+  | {
+      type: 'affine:list' | 'affine:code';
+      text: string;
+    }
+  | {
+      type: 'affine:image';
+      blobId: string;
+      alt: string;
+    };
+
+export interface PageDocContent {
+  title: string;
+  summary: string;
+  content: PageDocContentType[];
+}
+
+export interface WorkspaceDocContent {
+  name: string;
+  avatarKey: string;
+}
+
+type KnownFlavour =
+  | 'affine:page'
+  | 'affine:paragraph'
+  | 'affine:list'
+  | 'affine:code'
+  | 'affine:image';
+
+export function parseWorkspaceDoc(doc: Doc): WorkspaceDocContent | null {
+  // not a workspace doc
+  if (!doc.share.has('meta')) {
+    return null;
+  }
+
+  const meta = doc.getMap('meta');
+
+  return {
+    name: meta.get('name') as string,
+    avatarKey: meta.get('avatar') as string,
+  };
+}
+
+export interface ParsePageOptions {
+  maxSummaryLength: number;
+}
+
+export function parsePageDoc(
+  doc: Doc,
+  opts: ParsePageOptions = { maxSummaryLength: 150 }
+): PageDocContent | null {
+  // not a page doc
+  if (!doc.share.has('blocks')) {
+    return null;
+  }
+
+  const blocks = doc.getMap<Map<any>>('blocks');
```

Comment:
The contents of blocks do not necessarily appear in order. Obviously, the order of the blocks is sensitive here, and simply traversing the blocks is not enough. You should start from the root block (`flavour` === `affine:page`) and traverse the tree structure.

![CleanShot 2024-08-09 at 10 22 01](https://github.com/user-attachments/assets/d1cd6cf9-c875-4086-af7e-385267fdf218)

Reply:
[![dunno](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/WjlTemxb6oA4PgZFmj08/70e4facb-a079-46c8-bbe4-e7101e43de8c/dunno.gif)](https://app.graphite.dev//settings/meme-library?org=toeverything)